### PR TITLE
[BugFix] Ensure SinkIOBuffer outlives execution queue

### PR DESCRIPTION
**Pull request by letian-jiang**
Fix [SinkIOBuffer use-after-free in execution queue caused by pending task](https://github.com/StarRocks/StarRocksTest/issues/2691)

## What type of PR is this:
- [x] BugFix
- [ ] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Checklist:
- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.1
  - [x] 3.0
  - [x] 2.5
  - [ ] 2.4

### CODE REVIEW

File: be/src/exec/pipeline/sink/sink_io_buffer.h [hunk 1]
```diff
@@ -84,10 +84,12 @@ class SinkIOBuffer {
     virtual void cancel_one_sinker() { _is_cancelled = true; }
 
     virtual void close(RuntimeState* state) {
-        _is_finished = true;
         if (_exec_queue_id != nullptr) {
             bthread::execution_queue_stop(*_exec_queue_id);
+            bthread::execution_queue_join(*_exec_queue_id); // make sure the stop task has been executed
```
Comment (mofeiatwork) [inline be/src/exec/pipeline/sink/sink_io_buffer.h:89]: had better to handle the return code of these two function `stop` and `join`

**Comment by silverbullet233** [comment]
> LGTM. How do you write UT to reproduce this case?

It is a little difficult to reproduce and very hard to write UT

File: be/src/exec/pipeline/sink/sink_io_buffer.h (hunk 1 above)
Comment (letian-jiang) [reply be/src/exec/pipeline/sink/sink_io_buffer.h]: added error logging

**Review by murphyatwork — APPROVED** [review]

File: be/src/exec/pipeline/sink/sink_io_buffer.h [hunk 2]
```diff
@@ -84,10 +84,21 @@ class SinkIOBuffer {
...
-        _is_finished = true;
         if (_exec_queue_id != nullptr) {
-            bthread::execution_queue_stop(*_exec_queue_id);
+            int ret = bthread::execution_queue_stop(*_exec_queue_id);
+            if (ret != 0) {
+                auto error_msg = fmt::format("Fail to stop execution queue: {}", std::strerror(errno));
+                LOG(WARNING) << error_msg;
+            }
+
+            ret = bthread::execution_queue_join(*_exec_queue_id); // make sure the stop task has been executed
```
Comment (silverbullet233) [inline be/src/exec/pipeline/sink/sink_io_buffer.h:94]: will this interface block and wait until it finishes?

Comment (letian-jiang) [reply be/src/exec/pipeline/sink/sink_io_buffer.h:94]: yes

**Review by Youngwb — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
@Mergifyio backport branch-3.1

**Comment by letian-jiang** [comment]
This PR is reverted because of a deadlock problem. 

I will open another PR to resolve this issue.

**Pull request merged**